foreman/data_refinery_foreman/foreman/main.py

```diff
@@ -0,0 +1,260 @@
+    """If a job fails too many times, log it and stop retrying."""
+    # Not strictly retried but will prevent the job from getting
+    # retried any more times.
+    job.retried = True
+
+    # success may already be False, but if it was a hung or lost job
+    # this will ensure it's marked as failed.
+    job.success = False
+    job.save()
+
+    # At some point this should become more noisy/attention
+    # grabbing. However for the time just logging should be sufficient
+    # because all log messages will be closely monitored during early
+    # testing stages.
+    logger.warn("%s #%d failed %d times!!!", job.__class__.__name__, job.id, MAX_NUM_RETRIES + 1)
+
+
+def handle_downloader_jobs(jobs: List[DownloaderJob]) -> None:
+    """For each job in jobs, either retry it or log it."""
+    for job in jobs:
+        if job.num_retries < MAX_NUM_RETRIES:
+            requeue_downloader_job(job)
+        else:
+            handle_repeated_failure(job)
+
+
+def do_forever(min_loop_time: timedelta) -> Callable:
+    """Run the wrapped function in a loop forever.
+
+    The function won't be run more often than once per min_loop_time,
+    however if it takes longer to run than min_loop_time, then it will
+    be run less often than once per min_loop_time.
+    """
+    def decorator(function: Callable) -> Callable:
+        @wraps(function)
+        def wrapper(*args, **kwargs):
+            while(True):
+                start_time = timezone.now()
+
+                function(*args, **kwargs)
+
+                loop_time = timezone.now() - start_time
+                if loop_time < min_loop_time:
+                    remaining_time = MIN_LOOP_TIME - loop_time
+                    time.sleep(remaining_time.seconds)
+
+        return wrapper
+    return decorator
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_failed_downloader_jobs() -> None:
+    """Handle downloader jobs that were marked as a failure."""
+    failed_jobs = DownloaderJob.objects.filter(success=False, retried=False)
+    handle_downloader_jobs(failed_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_hung_downloader_jobs() -> None:
+    """Retry downloader jobs that were started but never finished."""
+    minimum_start_time = timezone.now() - MAX_DOWNLOADER_RUN_TIME
+    hung_jobs = DownloaderJob.objects.filter(
+        success=None,
+        retried=False,
+        end_time=None,
+        start_time__lt=minimum_start_time
+    )
+
+    handle_downloader_jobs(hung_jobs)
+
+
+@do_forever(MIN_LOOP_TIME)
+def retry_lost_downloader_jobs() -> None:
+    """Retry downloader jobs that went too long without being started.
+
+    Idea: at some point this function could integrate with the spot
+    instances to determine if jobs are hanging due to a lack of
+    instances. A naive time-based implementation like this could end
+    up retrying every single queued job if there were a long period of
+    spot instance bid price being very high.
```

Comment:
I'll do it here since that comment is mostly a note to myself anyway. (Unless of course you'd like me to add it in the comment.)

Basically we're planning on having the Data Refinery use AWS spot instances (or the google/microsoft equivalent). The way those work is that you set a bid price, and when the price of the spot instances drops below your bid price you get however many machines you specified you wanted until the price rises above your bid price.

This comment is referring to a scenario where the spot instance price is higher than our bid price for long enough that a job got queued by the surveyor (which runs on a persistent instance) and then sat in the SQS queue long enough that the foreman thought it somehow got lost forever (since there were no worker instances to pick it up). There's probably a number of ways this scenario could be handled, this comment is mostly a reminder that once I actually use spot instances that there could be an issue with jobs getting "lost" when actually they just haven't had a chance to get picked up yet.